### PR TITLE
refacting: crdt items refs optimize

### PR DESCRIPTION
**Pull request by forehalo**
replace `Arc` with `Somr` (Single owner multiple references lazily dropped head data type)

### CODE REVIEW

File: libs/jwst-codec/src/doc/codec/mod.rs [hunk 1]
```diff
@@ -10,12 +10,12 @@
...
 
 pub use any::Any;
-pub use content::Content;
+pub(crate) use content::Content;
 pub use delete_set::DeleteSet;
 pub use id::{Client, Clock, Id};
 pub use io::{CrdtRead, CrdtReader, CrdtWrite, CrdtWriter, RawDecoder, RawEncoder};
-pub use item::{item_flags, Item, ItemFlags, ItemRef, Parent};
-pub use refs::StructInfo;
+pub(crate) use item::{item_flags, Item, ItemFlags, ItemRef, Parent};
```
[the remark this answers is not in the document]
Comment (Brooooooklyn) [reply libs/jwst-codec/src/doc/codec/mod.rs:17]: remove `item_flags` import

File: libs/jwst-codec/src/doc/codec/content.rs [hunk 2]
```diff
@@ -212,36 +203,25 @@ impl Content {
...
-
-                match ty.kind {
-                    YTypeKind::XMLElement | YTypeKind::XMLHook => {
-                        encoder.write_var_string(ty.name.as_ref().unwrap())?;
+                if let Some(ty) = ty.get() {
+                    let ty = ty.read().unwrap();
+                    let type_ref = u64::from(ty.kind());
+                    encoder.write_var_u64(type_ref)?;
+
+                    match ty.kind {
```
Comment (Brooooooklyn) [inline libs/jwst-codec/src/doc/codec/content.rs:219]: `if matches!(ty.kind, YTypeKind::XMLElement | YTypeKind::XMLHook) {} `

File: libs/jwst-codec/src/doc/codec/content.rs [hunk 3]
```diff
@@ -212,36 +203,25 @@ impl Content {
             Self::String(string) => {
                 encoder.write_var_string(string)?;
             }
-            Self::Embed(json) => {
-                encoder.write_var_string(json.to_string())?;
+            Self::Embed(val) => {
+                encoder.write_var_string(serde_json::to_string(val).unwrap())?;
```
Comment (Brooooooklyn) [inline libs/jwst-codec/src/doc/codec/content.rs:207]: Don't `unwrap` unless absolutely necessary, you can use `map_err` here or implement `From<serde_json::Error>` for `JwstError`.

File: libs/jwst-codec/src/doc/codec/content.rs [hunk 4]
```diff
@@ -212,36 +203,25 @@ impl Content {
...
             }
             Self::Format { key, value } => {
                 encoder.write_var_string(key)?;
-                encoder.write_var_string(value.to_string())?;
+                encoder.write_var_string(serde_json::to_string(value).unwrap())?;
```
Comment (Brooooooklyn) [inline libs/jwst-codec/src/doc/codec/content.rs:211]: same here

File: yarn.lock [hunk 5]
```diff
@@ -0,0 +1,1176 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (darkskygit) [inline yarn.lock:1]: redundant lock file

File: libs/jwst-codec/src/doc/document.rs [hunk 6]
```diff
@@ -312,49 +312,44 @@ mod tests {
     }
 
     #[test]
+    #[cfg_attr(any(miri, loom), ignore)]
```
Comment (darkskygit) [inline libs/jwst-codec/src/doc/document.rs:315]: cannot pass loom test?

Comment (forehalo) [reply libs/jwst-codec/src/doc/document.rs:315]: loom always reports "Model exceeded maximum number of branches" at `array.insert(idx, sub_arr)`, will look on it later on

**Review by darkskygit — APPROVED** [review]

**Pull request merged**